### PR TITLE
PSR2/ClassDeclaration: support namespace relative names / prevent fixer conflict

### PR DESCRIPTION
**Pull request by jrfnl**
### PSR2/ClassDeclaration: prevent fixer conflict with itself [1]

While the `PSR2.Classes.ClassDeclaration` sniff did take partially/fully qualified names into account for interfaces being extended, it did not take _namespace relative_ interface name into account.

This led to a fixer conflict within the sniff, where the sniff would first add a space between the `namespace` keyword and the namespace separator (`SpaceBeforeName` fixer) and in a subsequent loop would remove that same space again as it would think it was a space before a comma (`SpaceBeforeComma` fixer).

Fixed now by adding support for namespace relative interface names in the `extends` checks.

Includes unit test.

### PSR2/ClassDeclaration: prevent fixer conflict with itself [2]

While the `PSR2.Classes.ClassDeclaration` sniff did take partially/fully qualified names into account for interfaces being implemented, it did not take _namespace relative_ interface names into account.

This led to a fixer conflict within the sniff, where the sniff would first add a newline between the `namespace` keyword and the namespace separator (`InterfaceSameLine` fixer) and in a subsequent loop would remove that same new line again as it would think it was a space before a comma (`SpaceBeforeComma` fixer).

Fixed now by adding support for namespace relative interface names in the `implements` check.

Includes unit test.

---

Note: at a glance, this sniff could probably do with a more thorough review and more defensive token walking/more precise checking, but that's outside the scope of this PR.

## Suggested changelog entry
PSR2.Classes.ClassDeclaration: using namespace relative interface names in the extends/implements part of a class declaration would lead to a fixer conflict.

## Related issues/external references

Related to #152

## Types of changes
- [x] Bug fix _(non-breaking change which fixes an issue)_

### CODE REVIEW

File: src/Standards/PSR2/Sniffs/Classes/ClassDeclarationSniff.php [hunk 1]
```diff
@@ -314,12 +314,14 @@ public function processOpen(File $phpcsFile, $stackPtr)
...
                 && ($tokens[($className - 1)]['code'] !== T_NS_SEPARATOR
-                || $tokens[($className - 2)]['code'] !== T_STRING)
+                || ($tokens[($className - 2)]['code'] !== T_STRING
+                && $tokens[($className - 2)]['code'] !== T_NAMESPACE))
             ) {
                 $prev = $phpcsFile->findPrevious(
                     [
                         T_NS_SEPARATOR,
                         T_WHITESPACE,
+                        T_NAMESPACE,
```
Comment (fredden) [inline src/Standards/PSR2/Sniffs/Classes/ClassDeclarationSniff.php:324]: Is this change necessary? The code seems to behave the same way with and without this line.

**Review by fredden — COMMENTED** [review]
I have tested this locally and can confirm that the code changes cover the added tests appropriately, and that the old code reported "FAILED TO FIX" for these test cases.

I have a query about one change, which we can address in the thread below.

File: src/Standards/PSR2/Sniffs/Classes/ClassDeclarationSniff.php (hunk 1 above)
Comment (jrfnl) [reply src/Standards/PSR2/Sniffs/Classes/ClassDeclarationSniff.php:324]: You are right, it is not necessary. I did it for consistency more than anything else. Removing it now.

I will leave this for a proper sniff review as per:

> Note: at a glance, this sniff could probably do with a more thorough review and more defensive token walking/more precise checking, but that's outside the scope of this PR.

For the record: the above note was triggered by seeing code like line 451-453 which checks that the code _after_ a name is not a `\` and not a comma and then presumes it is whitespace, but doesn't actually check if the token is whitespace (and replaces the token anyhow, potentially destroying valid code).

This will need further investigation (and proof of the bug via a code sample, not just my experience telling me there is a likely to be bug). It might be that the code earlier on in the sniff already checks for `\`, comma and whitespace, so that if the next token is not `\` or comma, the sniff can be sure that it will be whitespace (but then again, why not check for that in line 451-453 ?). Investigating that was outside the scope of the current PR though.
Considering the bugs now fixed and seeing that code, I have a feeling there will be more, potentially incorrect, presumptions build into the sniff, which is the reason why I added that note.

Comment (jrfnl) [reply src/Standards/PSR2/Sniffs/Classes/ClassDeclarationSniff.php]: I've opened ticket #425 to prevent these notes about the sniff review being lost once this PR has been merged.

**Review by fredden — APPROVED** [review]

**Comment by jrfnl** [comment]
Rebased without changes, merging once the build has passed.
